File: Python/1D Parity/1D_Parity.py
import functools, random

def parityGenerator(msg):
    binary_li = list(map(int, list(msg)))
    parity = functools.reduce(lambda a, b: a ^ b, binary_li)
    return parity

def parityChecker(msg):
    p = parityGenerator(msg)
    if p == 0:
        return True
    else:
        return False

def sender(msg):
    return msg + str(parityGenerator(msg))

def transmission(msg):
    li = list(map(int, list(msg)))
    if random.randint(0, 1):
        error_pos = random.randint(0, len(msg)-1)
        li[error_pos] = 1 - li[error_pos]
    return ''.join(list(map(str, li)))
        
def receiver(msg):
    return parityChecker(msg), msg[:-1]



data = input('Enter data to send: ')
msg_send = sender(data)
sent = transmission(msg_send)
p, msg_received = receiver(sent)

if p:
    print('The transmission was ERROR-FREE!\nThe data received is: ' + msg_received)
else:
    print('There was an ERROR in the data received!\nThe data received is: ' + msg_received)

File: Python/1D Parity/1D_Parity.c
#include <stdio.h>
#include <malloc.h>
#include <string.h>

char parity(char num[]){
    int size = strlen(num);
    int count = 0;
    for (int i = 0; i < size; i++){
        if(num[i] == '1')
            count++;
    }

    if(count%2 == 0)
        return '0';
    else
        return '1';
}

char* sender(char data[]){
    char *d = data;
    int size = strlen(d);
    d[size] = parity(d);
    d[size+1] = '\0';

    return d;
}

char* reciever(char data[]){
    char *d = data;
    int size = strlen(d);

    if(parity(d) == '0')
        printf("\nTransmission was ERROR-FREE\n");
    else
        printf("\nThere was an ERROR in the transmission\n");

    d[size-1] = '\0';
    return d;
}

int main(int argc, char const *argv[])
{
    char num[128];
    printf("Enter data: ");
    scanf("%[^\n]s", &num);

    char* sent = sender(num);
    printf("Data sent, after appending the parity bit: %s", sent);

    char* received = reciever(sent);
    printf("Data recieved: %s\n\n", sent);
    
    return 0;
}
